yara-x-dump/src/serializer.rs

```diff
@@ -0,0 +1,573 @@
+use chrono::prelude::{DateTime, NaiveDateTime, Utc};
+use protobuf::descriptor::FieldDescriptorProto;
+use protobuf::reflect::MessageRef;
+use protobuf::reflect::ReflectFieldRef;
+use protobuf::reflect::ReflectValueRef;
+use protobuf_support::text_format::quote_bytes_to;
+use std::fmt::Write;
+use yansi::Color;
+
+use crate::dumper::exts::field_options;
+
+use crate::Error;
+
+// A struct that represents serializers for different formats
+struct JsonSerializer;
+struct YamlSerializer;
+struct TomlSerializer;
+struct XmlSerializer;
+
+// A struct that represents colors for output
+struct Colors;
+
+impl Colors {
+    const GREEN: Color = Color::RGB(51, 255, 153);
+    const BLUE: Color = Color::RGB(51, 51, 255);
+    const YELLOW: Color = Color::RGB(255, 255, 102);
+}
+
+// A struct that represents options for a field values
+#[derive(Debug, Default, Clone)]
+struct ValueOptions {
+    is_hex: bool,
+    is_timestamp: bool,
+}
+
+/// A trait for any type that can serialize a message
+pub(crate) trait Serializer {
+    /// Serialize a message
+    ///
+    /// # Arguments
+    ///
+    /// * `message`: The message to serialize
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result<String, Error>` where the `String` is the serialized
+    /// message in specified format and the `Error` is any error that occurred
+    /// during serialization
+    ///
+    /// # Errors
+    ///
+    /// * `Error::ParsingJSONError`: If the message is not a valid JSON
+    /// * `Error::ParsingYAMLError`: If the message is not a valid YAML
+    /// * `Error::ParsingTOMLError`: If the message is not a valid TOML
+    /// * `Error::ParsingXMLError`: If the message is not a valid XML
+    fn serialize(&self, message: &str) -> Result<String, Error>;
+}
+
+/// Implement the trait for the JSON serializer
+impl Serializer for JsonSerializer {
+    fn serialize(&self, message: &str) -> Result<String, Error> {
```

Comment:
`JsonSerializer ` receives a JSON-formatted string and returns a JSON-formatted string, but with pretty print. `YamlSerializer ` receives a JSON-formatted string and produces the same struct as a YAML-formatted string.
`TomlSerializer ` receives a JSON-formatted string and produces the same struct as a TOML-formatted string.

So, the whole process is:

1. convert a `&dyn Message` to a JSON-formatted string.
2. use one implementation of `Serializer` to convert the JSON-formatted string to YAML, TOML, or pretty-printed JSON.

This process is unnecessarily inefficient as it requires converting the protobuf data to JSON, and from JSON to another format (one of which is also JSON!). We have a way to produce a JSON directly from a protobuf message using https://docs.rs/protobuf-json-mapping/latest/protobuf_json_mapping. I'm not sure it supports pretty-printing, but I would say that pretty JSON output is not our priority, as JSON can be prettified with external tools. XML is mostly dead, I doubt anyone will use XML when you have JSON.
And TOML is ok for simple formats and config files, but for very nested structures I don't think is very useful. Having JSON and YAML is enough.

The other format we want to support is YAML, and that's exactly what we need to build: a crate that takes a protobuf message and produces a YAML.

Reply:
I agree! At first I wanted to provide an unify interface to handle multiple types of serilization and therefore have multiple types of machine-readable output. This was a bit too much as I agree with the fact that YAML and JSON are enough. I have reworked the crate to specify only on YAML and JSON is done via built-in function provided by `protobuf-json-mapping` crate.

I have also implemented a color switch, that turns on colors for both JSON and YAML via CLI. YANSI colors are used and default is that they are disabled. For JSON this also provides pretty-print.